[17.01] Fix Conda --offline logic for bug in 4.0,4.1,4.3.

jmchilton opened the pull request:
xref conda/conda#2035

We haven't ever explicitly targeted these versions - but people seem to have them frequently. There have been several reports of planemo problems for people that landed up with 4.3.

Ping @bgruening or @mvdbeek - does this look right?

jmchilton committed:
WIP: Fix for conda=4.3

jmchilton committed:
Adjust offline fix to work for a broader range of potential target Condas.

This logic could in theory be backported all the way to 16.01 I suppose now.

nsoranzo commented on lib/galaxy/tools/deps/conda_util.py:
```diff
@@ -495,7 +528,14 @@ def build_isolated_environment(
+        offline_works = conda_context.conda_version < LooseVersion("4") or \
+            (conda_context.conda_version >= LooseVersion("4.2") and conda_context.conda_version < LooseVersion("4.3"))
+        if offline_works:
+            create_args.extend(["--offline"])
```
create_args.append("--offline")

nsoranzo commented on lib/galaxy/tools/deps/conda_util.py:
```diff
@@ -495,7 +528,14 @@ def build_isolated_environment(
+        if offline_works:
+            create_args.extend(["--offline"])
+        else:
+            create_args.extend(["--use-index-cache"])
```
create_args.append("--use-index-cache")

jmchilton replied:
I used extend intentionally - it reads better IMO with the rest of the code in that that little block (more uses of the ``extend`` function). Hope this is okay.

nsoranzo replied:
Fair enough!

bgruening commented:
I guess this is needed, sad, but needed. Can you please add a comment/todo that this should be removed at some point.

jmchilton commented:
@bgruening Why a comment about removing this? Won't these versions always be broken? We may need to restore it for 4.4 or some later revision at some point - but I can created an issue to track that.

bgruening commented:
True, lets create an issue for this.

mvdbeek commented:
Is this really broken in 4.0 and 4.1 ? The only time I saw this it happened in 4.3.

jmchilton commented:
@mvdbeek So I just tried 4.0 and 4.1 with --offline and they seemed to work fine. I have no clue why I was so confused. I thought maybe the --offline bug was what was preventing us from upgrading to 4.X for so long? I'll rework this.

nsoranzo commented on lib/galaxy/tools/deps/conda_util.py:
```diff
@@ -82,6 +82,39 @@ def __init__(self, conda_prefix=None, conda_exec=None,
+                package_parts = package.split("-", 2)
+                if len(package_parts) < 3:
+                    continue
+                package, version, build = package_parts
```
Since the package name can contain dashes, this may result in parsing e.g. `conda-build-2.0.11-py27_0.json` as `package = 'conda', version = 'build', build = '2.0.11-py27_0.json'` .

I think a better solution would be:
```python
                package_parts = package.split("-")
                if len(package_parts) < 3:
                    continue
                package = '-'.join(package_parts[:-2])
                version = package_parts[-2]
                build = package_parts[-1]
```

jmchilton committed:
More revisions to fixing --offline mode to Conda.

Incorporate package extraction fixes thanks to @nsoranzo, the fix for version 4.1 and 4.0 thanks to @mvdbeek, a comment for @bgruening and a new test tool that makes it a bit easier to test this.

jmchilton commented:
So I tried to incorporate everyones awesome suggestions. Thanks for the detailed review all!

mvdbeek commented on test/functional/tools/mulled_example_multi_incomplete.xml:
```diff
@@ -0,0 +1,19 @@
+    <requirements>
+        <requirement type="package" version="1.3.1">samtools</requirement>
+        <requirement type="package" version="2.26.0">bedtools</requirement>
+        <requirement type="package" version="4.5">notarealpackage</requirement>
```
I'm not sure I understand why notarealpackage is included. The offline bug seems to be independent of whether a dependency can be resolved, right?

jmchilton replied:
Without this the Conda dependencies are all evaluated together using "resolve_all" and so an environment is never created I don't think - at least not a per-job environment. I can update this to add a comment about this.

mvdbeek replied:
Of course, you're right!

jmchilton commented:
So this is why I was confused - I think offline was broken at some point. Maybe it just isn't in the latest 4.0 and 4.1 release? From @bgruening on that Conda issue:

```
I end up in a RuntimeError: maximum recursion depth exceeded error with the --offline argument.

The problem exists since conda 4.0. I was able to get it running with "conda install conda=3.19.0"
```

This seems like a sticky issue - we don't need to hold the release up on this I don't think.

jmchilton committed:
Add a clarification to new test tool.

mvdbeek commented:
Thanks for taking care of this @jmchilton !
